cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1909,6 +1909,168 @@ class SelectKUnstableMetaFunction : public MetaFunction {
+
+    NullPartitionResult sorted =
+        array_sorter(sort_begin, sort_end, arr, 0, array_options);
+    uint64_t rank = 0;
+
+    ARROW_ASSIGN_OR_RAISE(auto rankings,
+                          MakeMutableUInt64Array(uint64(), length, ctx_->memory_pool()));
+    auto out_begin = rankings->GetMutableValues<uint64_t>(1);
+
+    switch (tiebreaker_) {
+      case RankOptions::Dense: {
+        T currValue, prevValue;
+        for (auto it = sorted.overall_begin(); it < sorted.overall_end(); it++) {
+          currValue = GetView::LogicalValue(arr.GetView(*it));
+          if (rank == 0 || it == sorted.nulls_begin || it == sorted.nulls_end ||
+              currValue != prevValue) {
+            rank++;
+          }
+
+          out_begin[*it] = rank;
+          prevValue = currValue;
+        }
+        break;
+      }
+      case RankOptions::First: {
+        for (auto it = sorted.overall_begin(); it < sorted.overall_end(); it++) {
+          out_begin[*it] = ++rank;
+        }
+        break;
+      }
+      case RankOptions::Min: {
+        T currValue, prevValue;
+        for (auto it = sorted.overall_begin(); it < sorted.overall_end(); it++) {
+          currValue = GetView::LogicalValue(arr.GetView(*it));
+          if (rank == 0 || it == sorted.nulls_begin || it == sorted.nulls_end ||
+              currValue != prevValue) {
+            rank = (it - sorted.overall_begin()) + 1;
+          }
+          out_begin[*it] = rank;
+          prevValue = currValue;
+        }
+        break;
+      }
+      case RankOptions::Max: {
+        T currValue, prevValue;
+        auto rank = sorted.overall_end() - sorted.overall_begin();
+        for (auto it = sorted.overall_end() - 1; it >= sorted.overall_begin(); it--) {
+          currValue = GetView::LogicalValue(arr.GetView(*it));
+          if ((it < sorted.overall_end() - 1 && (currValue != prevValue)) ||
+              it == sorted.nulls_begin - 1 || it == sorted.nulls_end - 1) {
+            rank = it - sorted.overall_begin() + 1;
+          }
+          out_begin[*it] = rank;
+          prevValue = currValue;
+        }
+        break;
+      }
+    }
+
+    *output_ = Datum(rankings);
+    return Status::OK();
+  }
+
+  ExecContext* ctx_;
+  const Array& array_;
+  SortOrder order_;
+  NullPlacement null_placement_;
+  RankOptions::Tiebreaker tiebreaker_;
+  const std::shared_ptr<DataType> physical_type_;
+  Datum* output_;
+};
+
+const FunctionDoc rank_doc(
+    "Returns the ranking of an array",
+    ("This function computes a rank of the input array.\n"
+     "By default, null values are considered greater than any other value and\n"
+     "are therefore sorted at the end of the input. For floating-point types,\n"
+     "NaNs are considered greater than any other non-null value, but smaller\n"
+     "than null values. The default tiebreaker is to assign ranks in order of\n"
+     "when ties appear in the input\n"
```

Comment:
Nit: add stop at end of sentence.
```suggestion
     "than null values. The default tiebreaker is to assign ranks in order of\n"
     "when ties appear in the input.\n"
```